build/bazel/remote/execution/v2/remote_execution.proto

```diff
@@ -1961,9 +1961,30 @@ message Compressor {
 
     // Brotli compression.
     BROTLI = 3;
+
+    // Zstandard compression with dictionary.
+    //
+    // When this is used, the server MUST advertise the dictionaries by
+    // including
+    // [ZstdDictionaryRegistry][build.bazel.remote.execution.v2.ZstdDictionaryRegistry]
+    // digest in CacheCapabilities.
+    ZSTD_DICT = 4;
   }
 }
 
+message ZstdDictionaryRegistry {
+  // Each file respresent a single Zstandard dictionary
+  // with name being the `dictId` that was added to the header
+  // of the compressed file.
+  //
+  // A special file with name `default` is used to represent the
+  // default dictionary that is used when no `dictId` is specified.
+  // Clients SHOULD prefer to use the default dictionary when possible.
+  //
+  // The `default` dictionary MUST be present.
+  repeated FileNode dictionaries = 1;
```

Comment:
As I mentioned during our monthly meeting, I don't see why we must shoehorn this into `Directory`/`FileNode`/... related messages. We're trying to store properties about compression dictionaries. Not a file system hierarchy. There is never going to be a scenario where a build action wants to call `ls` against a set of compressor dictionaries. Dictionaries also don't need an `is_executable` or `node_properties`.

Why permit the use case where no dictionary ID is present? Isn't that just encouraging our users to paint themselves into a corner? For what is a brand new feature I see no reason to be so liberal here.

Please do something like this instead.
```proto
message ZstdDictionaryRegistry {
  // The dictionary ID that clients SHOULD use when uploading
  // objects into the Content Addressable Storage. An entry for
  // this dictionary ID MUST be present in `dictionaries`.
  uint32 upload_dictionary_id = 1;

  // Dictionaries that may be used by the server when returning
  // objects stored in the Content Addressable Storage. The key
  // corresponds to a dictionary ID, as described in RFC 8878,
  // section 3.1.1.1.3, while the value refers to an dictionary
  // as described in RFC 8878, chapter 6.
  map<uint32, Digest> dictionaries = 2;
}
```

A downside of `map<X, Y>` is that Protobuf libraries tend to be inconsistent in the order in which they get marshalled. I can imagine that may be annoying for us, as they get stored in a CAS. If that's a concern, we can use a `repeated` field instead.

Reply:
I think I would prefer a repeated field and an index-into-that field.

Reply:
I was fixated on FileNode initially as I want to give more flexibility in naming the dictionary for other compression algorithms. However since then, I have decided to make this specifically for Zstandard dictionary compression so I think using `uint32` is a great suggestion here.

```proto
message ZstdDictionaryRegistry {
  message Dictionary {
    uint32 id 1;
    Digest digest 2;
  }

  // The dictionary ID that clients SHOULD use when uploading
  // objects into the Content Addressable Storage. An entry for
  // this dictionary ID MUST be present in `dictionaries`.
  uint32 default_dictionary_id = 1;

  // Dictionaries that may be used by the server when returning
  // objects stored in the Content Addressable Storage. The key
  // corresponds to a dictionary ID, as described in RFC 8878,
  // section 5, while the value refers to a dictionary
  // as described in RFC 8878, chapter 6.
  repeated Dictionary dictionaries = 2;
}
```

Would something like this make sense?

Note: I do want to `s/upload/default/` to make sure that we generalize the use of a default dictionary for unforeseeable use cases.

I also think the dictionary ID is described in section 5, not 3.1.1.3 https://datatracker.ietf.org/doc/html/rfc8878#name-dictionary-format